### PR TITLE
Clipboard History

### PR DESCRIPTION
**Pull request by wise0704**
image
image

### CODE REVIEW

**Comment by wise0704** [comment]
An alternative to the listbox texts is to wrap the `ClipboardTextData` in a class, and override `ToString()`
```c#
if (data.IsTextFormat)
{
    int lineBreak = data.Text.IndexOfAny(new[] { '\r', '\n' });
    return "[" + data.Format + "]" + (lineBreak >= 0 ? data.Text.Remove(lineBreak) : data.Text);
}
return "[" + data.Format + "]";
```

File: PluginCore/PluginCore/Resources/en_US.resx [hunk 1]
```diff
@@ -6169,4 +6169,8 @@ AND - All conditions are met.</value>
     <value>Paste History</value>
     <comment>Added after 5.2.0</comment>
   </data>
+  <data name="FlashDevelop.Description.ClipboardHistorySize" xml:space="preserve">
+    <value>The number of items to store in the clipboard history. Minimun value is 1.</value>
```
Comment (wise0704) [inline PluginCore/PluginCore/Resources/en_US.resx]: "minimu**n**"...

**Comment by wise0704** [comment]
Also, `ClipboardHistoryDialog` can derive `PluginCore.Controls.SmartForm` instead, then it will save its window size to a file (Settings\FormStates\\%FormGuid%.fdb).
But it will also do probably unwanted things like `ScaleHelper.AdjustForHighDPI()`.

**Comment by Neverbirth** [comment]
Some random thoughts:

- I'm a bit concerned about using a hook. Starting from Windows Vista MS added some protections that may result in applications being removed from a global hook like this, and you won't know about it, It used to happen in an application from my previous company; there is a registry entry for this, although it requires admin rights and would feel hacky and bad for us having to change it. I don't think it would happen in a case like this, because the protection is mainly about a program taking too long to process a message and pass it to the next listener in the chain, but who knows. 

- Why do you even put things that cannot be pasted into the list? I don't think it's worth it. But by your comments I understand this PR is still WIP.

I will try to play with this more tomorrow. I checked VS out of curiosity and it doesn't seem to use a global clipboard hook? I will try IJ tomorrow.

**Comment by wise0704** [comment]
- https://msdn.microsoft.com/en-us/library/ms649016(v=vs.85).aspx#dataxchg_Monitoring_Clipboard_Contents
  >There are three ways of monitoring changes to the clipboard. The oldest method is to create a clipboard viewer window. Windows 2000 added the ability to query the clipboard sequence number, and Windows Vista added support for clipboard format listeners. Clipboard viewer windows are supported for backward compatibility with earlier versions of Windows. New programs should use clipboard format listeners or the clipboard sequence number.

  I think the hook you're referring to is [SetClipboardViewer](https://msdn.microsoft.com/en-us/library/ms649052(v=vs.85).aspx), which is the window handle chain method. The one I'm using is [AddClipboardFormatListener](https://msdn.microsoft.com/en-us/library/ms649033(v=vs.85).aspx):

  >This method is recommended over creating a clipboard viewer window because it is simpler to implement and avoids problems if programs fail to maintain the clipboard viewer chain properly or if a window in the clipboard viewer chain stops responding to messages.

- I was thinking of letting users re-copy contents from the history, back into the clipboard to allow users the "restore previous clipboard data" feature. 
That wasn't implemented yet mainly because I couldn't be bothered with inserting bitmap as RTF in the rich text box (users need to at least see the preview of the content they're about to restore).
But as you said, non-textual data can just be ignored. I'll leave the decision to you.

- I'd say, if possible, limiting the clipboard history to only the copy operations from scintilla control is not suitable. For example, you could want to copy&paste a file name from the project panel. A normal Ctrl+V would paste the content, but if it doesn't appear in the history, it wouldn't make too much sense.

**Comment by Neverbirth** [comment]
1. Could be, I will have to check this in depth. In the meantime I tried IJ, IJ is using a global clipboard checker for its history.
2. I don't get what you mean with letting users re-copy back into the history :S. From my side I would just display and paste RTF text in plain format, not RTF... it will be a pain in the ass for some stuff, also the RichText control wrapped by Windows Forms is an old one, you can change it to a newer one, but even so I wouldn't go that way. IJ is only capturing normal text, not even RTF.
3. VS is doing it that way heh, Although again, not IJ. Personally I don't have a big preference for one or the other.

**Comment by wise0704** [comment]
For 2., okay, I'll display only textual data in the history.

File: FlashDevelop/Dialogs/ClipboardHistoryDialog.cs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+            if (selectedData != null)
+            {
+                if (selectedData.Format == DataFormats.Text)
+                {
+                    richTextBox.Text = selectedData.Text;
+                }
+                else if (selectedData.Format == DataFormats.Rtf)
+                {
+                    richTextBox.Rtf = selectedData.Rtf;
```
Comment (Neverbirth) [inline FlashDevelop/Dialogs/ClipboardHistoryDialog.cs]: If I understood correctly you are not copying RTF anymore, so I guess this if is not needed.

File: FlashDevelop/Dialogs/ClipboardHistoryDialog.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+        #endregion
+
+        #region Events
+
+        private void ListBox_DrawItem(object sender, DrawItemEventArgs e)
+        {
+            e.DrawBackground();
+
+            string text = (e.Index + 1) + "    " + listBox.GetItemText(listBox.Items[e.Index]).TrimStart();
```
Comment (wise0704) [inline FlashDevelop/Dialogs/ClipboardHistoryDialog.cs]: Currently the ListBox only displays the first line:
image
If I change the code to
```c#
string[] lines = listBox.GetItemText(listBox.Items[e.Index]).Split('\n');
for (int i = 0; i < lines.Length; i++)
{
    lines[i] = lines[i].Trim();
}
string text = (e.Index + 1) + "    " + string.Join(" ", lines);
```
then it merges all lines together:
image

Which would be better?

**Pull request merged**